Changelog overlay

HoutarouOreki opened the pull request:
Mostly resolves #2886

Ready for reviews.

- [x] Header
- [x] Update Stream badges
- [x] ~Chart ([API works](https://streamable.com/1ger7))~
- [x] Changelog
- [x] Per-update stream changelog

[]()
- [x] ~Use no-break-space for `$" by {entry.GithubUser.DisplayName}"`~ not working
- [x] Match web sizing

ppy takeover
------------

- [x] Depends on https://github.com/ppy/osu-resources/pull/56.
- [x] Depends on #4830.
- [x] Depends on #4831.
- [x] Depends on #4832.
- [x] Depends on #4833.
- [x] Depends on #4834.
- [x] Depends on #4835.
- [x] Depends on #4837.
- [x] Depends on #4844.
- [x] Depends on ppy/osu-framework#2430.
- [x] Depends on ppy/osu-framework#2431.

---

HoutarouOreki committed:
Header1

HoutarouOreki committed:
Header2

HoutarouOreki committed:
Header3

HoutarouOreki committed:
Streams1

HoutarouOreki committed:
Streams2

HoutarouOreki committed:
Streams3

HoutarouOreki committed:
Refactor1, UX1

HoutarouOreki committed:
API1

HoutarouOreki committed:
Refactor2

AlFasGD commented on osu.Game/Graphics/UserInterface/TooltipIconButton.cs:
```diff
@@ -0,0 +1,43 @@
+        public TooltipIconButton()
+        {
+            Children = new Drawable[]
+                {
```
Decrease indent

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+                        SelectedRelease = streamBadge.ChangelogEntry;
+                        foreach (StreamBadge item in BadgesContainer.Children)
+                        {
+                            if (item.ChangelogEntry.Id != streamBadge.ChangelogEntry.Id) item.Deactivate();
```
Too long statement; move to new line after the condition

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+                    {
+                        SelectedRelease = streamBadge.ChangelogEntry;
+                        foreach (StreamBadge item in BadgesContainer.Children)
+                        {
```
Brackets can be safely removed

AlFasGD commented on osu.Game/Overlays/Changelog/Header/TextBadgePair.cs:
```diff
@@ -0,0 +1,125 @@
+        public void ShowText(double duration = 0, string displayText = null, Easing easing = Easing.InOutCubic)
+        {
+            LineBadge.IsCollapsed = false;
+            if (!string.IsNullOrEmpty(displayText)) Text.Text = displayText;
```
Use a new line after the condition

jorolf commented on osu.Game/Online/API/Requests/Responses/APIChangelog.cs:
```diff
@@ -0,0 +1,54 @@
+        public DateTimeOffset CreatedAt { get; set; }
+
+        [JsonProperty("disqus_id")]
+        public string DisqusId { get; set; }
```
I doubt we'll ever support disqus inside the client

AlFasGD commented on osu.Game/Overlays/Changelog/Header/TextBadgePairListing.cs:
```diff
@@ -0,0 +1,72 @@
+
+        protected override void OnHoverLost(InputState state)
+        {
+            if (IsActivated == false) LineBadge.ResizeHeightTo(1, LineBadge.TransitionDuration, Easing.Out);
```
Use `if (!IsActivated)` and new line after condition

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+                if (SelectedRelease != null)
+                {
+                    if (SelectedRelease.UpdateStream.Id != streamBadge.ChangelogEntry.Id)
+                    {
```
Brackets can be removed

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+                        streamBadge.Deactivate();
+                    }
+                }
+                else streamBadge.Deactivate();
```
Move statement to new line

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+        {
+            if (SelectedRelease == null)
+            {
+                foreach (StreamBadge streamBadge in BadgesContainer.Children) streamBadge.Activate(true);
```
Use new line between loop and statement

AlFasGD commented on osu.Game/Overlays/Changelog/ChangelogStreams.cs:
```diff
@@ -0,0 +1,98 @@
+        protected override void OnHoverLost(InputState state)
+        {
+            if (SelectedRelease == null)
+            {
```
Brackets can be safely removed

AlFasGD commented on osu.Game/Overlays/Changelog/Header/TextBadgePair.cs:
```diff
@@ -0,0 +1,125 @@
+
+        protected override bool OnHover(InputState state)
+        {
+            if (!IsActivated) sampleHover?.Play();
```
Use new line after condition

AlFasGD commented on osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            if (IsActivated)
+            {
+                if (displayText != Text.Text) ChangeText(transition_duration, displayText);
```
Use new line after condition

AlFasGD commented on osu.Game/Overlays/Changelog/Header/TextBadgePairRelease.cs:
```diff
@@ -0,0 +1,44 @@
+            else
+            {
+                ShowText(transition_duration, displayText);
+                IsActivated = true;
```
The value will always end up being true, so setting it outside won't be a problem

AlFasGD commented on osu.Game/Overlays/Changelog/StreamBadge.cs:
```diff
@@ -0,0 +1,133 @@
+                        new SpriteText
+                        {
+                            Text = ChangelogEntry.Users > 0 ?
+                                string.Join(" ", ChangelogEntry.Users.ToString("N0"), "users online"):
```
Prefer interpolated strings

AlFasGD commented on osu.Game/Overlays/Changelog/StreamBadge.cs:
```diff
@@ -0,0 +1,133 @@
+            isActivated = true;
+            this.FadeIn(transition_duration);
+            lineBadge.IsCollapsed = false;
+            if (!withoutHeaderUpdate) OnActivation?.Invoke();
```
Use new line after condition

AlFasGD commented on osu.Game/Overlays/ChangelogOverlay.cs:
```diff
@@ -0,0 +1,160 @@
+                    // the problem here is that when hovering over the builds' container
+                    // and pressing back, they don't lower their opacity they're rehovered on
+                    else header.ActivateListing();
+                    return true;
```
Can be removed

AlFasGD commented on osu.Game/Overlays/ChangelogOverlay.cs:
```diff
@@ -0,0 +1,160 @@
+            switch (action)
+            {
+                case GlobalAction.Back:
+                    if (header.IsListingActivated()) State = Visibility.Hidden;
```
Use new line after condition

AlFasGD commented on osu.Game/Overlays/ChangelogOverlay.cs:
```diff
@@ -0,0 +1,160 @@
+                content.Clear();
+                // should add listing to content here
+                if (!Streams.IsHovered)
+                    foreach (StreamBadge item in Streams.BadgesContainer.Children) item.Activate(true);
```
Use new line between loop and statement

AlFasGD commented on osu.Game/Overlays/ChangelogOverlay.cs:
```diff
@@ -0,0 +1,160 @@
+                if (!Streams.IsHovered)
+                    foreach (StreamBadge item in Streams.BadgesContainer.Children) item.Activate(true);
+                else
+                    foreach (StreamBadge item in Streams.BadgesContainer.Children) item.Deactivate();
```
Same as above

AlFasGD commented on osu.Game/Overlays/ChangelogOverlay.cs:
```diff
@@ -0,0 +1,160 @@
+            {
+                Streams.BadgesContainer.Clear();
+                foreach (APIChangelog item in res)
+                {
```
Brackets can be safely removed

HoutarouOreki committed:
Refactor3

peppy commented:
Please learn to split out and name your commits better. They should describe what you are doing inside them.

AlFasGD commented on osu.Game/Overlays/Changelog/StreamBadge.cs:
```diff
@@ -64,7 +64,7 @@ public StreamBadge(APIChangelog changelogEntry)
                         {
                             Text = ChangelogEntry.Users > 0 ?
-                                string.Join(" ", ChangelogEntry.Users.ToString("N0"), "users online"):
+                                string.Format($"{ChangelogEntry.Users:N0} users online") :
```
No need for `string.Format`, the interpolated string does the job by itself.

HoutarouOreki replied:
Appveyor doesn't seem happy without it
![image](https://user-images.githubusercontent.com/16261733/42993982-2aee4c12-8c0d-11e8-9ca8-b0e164ff3272.png)

HoutarouOreki replied:
Should I ignore it?

AlFasGD replied:
I have seen that error too, and I just specified the formatting culture, so using `.ToString("N0", CultureInfo.InvariantCulture)` is probably what it needs

HoutarouOreki replied:
InvariantCulture breaks formatting for other languages. Right now I have that number nicely formatted in my language, using InvariantCulture sets everything to English. Are we doing this?

HoutarouOreki committed:
Expand APIChangelog; Normalize its line endings

HoutarouOreki committed:
Add build-scoped requests;

Add OnClick to TooltipIconButton; Actions on pressing previous/next in builds

HoutarouOreki committed:
Remove accidental partial modifier in APIChangelog

HoutarouOreki committed:
Add neighboring builds handling;

+ provoke angry AppVeyor

HoutarouOreki committed:
Update comments; Improve neighboring builds handling;

Apply fixes to things pointed out by AppVeyor

HoutarouOreki committed:
Make chevrons updateable

HoutarouOreki committed:
Improve fake-api

HoutarouOreki committed:
Improvements to TooltipIconButton

Disable sounds when disabled; Remove default tooltip texts

HoutarouOreki committed:
Improve showing up builds

HoutarouOreki committed:
Unify build div's dimming mechanics with osu-web

HoutarouOreki committed:
Remove unnecessary statement; Uncapitalize private member;

HoutarouOreki committed:
Fix ChangelogEntries being a list of objects

HoutarouOreki committed:
Nullables in APIChangelog; Primitive changelog entries display

HoutarouOreki committed:
Improve primitive changelog entry formatting

HoutarouOreki committed:
Make build's creation date not nullable

HoutarouOreki committed:
Prevent scrolling to top on build change; Add bottom margin

HoutarouOreki committed:
Introduce better formatting to changelog entries; Sort by category

HoutarouOreki committed:
Change graph's colour on stream selection

HoutarouOreki committed:
Handle possible null reference;

HoutarouOreki committed:
Change chart's colour depending on release stream; Show whether it's populated

Fix chart requests

HoutarouOreki committed:
Check wether a chart is populated; Fixes to graph's colour setting

HoutarouOreki committed:
Show listing graph; Slight refactor

HoutarouOreki committed:
Show listing graph; Slight refactor

HoutarouOreki committed:
Set the right colour for background

HoutarouOreki committed:
Lower height of the cover image, Adjust height of the purple line

HoutarouOreki committed:
Add changelog listing

HoutarouOreki committed:
Use ClickableText for navigating to builds; Fix typo

HoutarouOreki committed:
Add ClickableText

HoutarouOreki committed:
ClickableText changes

HoutarouOreki committed:
Compress Test Case; Remove empty summary

HoutarouOreki committed:
Fix ClickableText test case

HoutarouOreki committed:
Use shorter expression

HoutarouOreki committed:
Merge branch 'master' into changelog-overlay

HoutarouOreki committed:
Merge branch 'changelog-overlay' of https://github.com/HoutarouOreki/osu into changelog-overlay

HoutarouOreki committed:
Use using

HoutarouOreki committed:
Show changelog by default

HoutarouOreki committed:
Place HTML messages in a TextFlowContainer

Add missing accessibility modifier in ClickableText test case and fix disposal

HoutarouOreki committed:
Fix ClickableText test case

HoutarouOreki committed:
Rewrite ChangelogOverlay.cs

HoutarouOreki committed:
Rewrite ChangelogHeader.cs

HoutarouOreki committed:
Rewrite ChangelogBadges

HoutarouOreki committed:
Rewrite ChangelogContent.cs

HoutarouOreki committed:
Add event handling in content groups

HoutarouOreki committed:
Strip test case

HoutarouOreki committed:
Rewrite improvements

HoutarouOreki committed:
Show on test case

HoutarouOreki committed:
Further rewrite

HoutarouOreki committed:
Refactor the rewrite

HoutarouOreki committed:
Merge branch 'master' into changelog-overlay

HoutarouOreki committed:
Follow up framework changes

HoutarouOreki committed:
Rewrite LineBadge.cs

+ update all its references

HoutarouOreki committed:
Change order in TextBadgePair.cs

HoutarouOreki committed:
Fix typos; remove outdated comments; minor order changes

HoutarouOreki committed:
Fix showing builds

HoutarouOreki committed:
Fix links not working due to rewrite regression

HoutarouOreki committed:
Handle around an APIChangelog instead of just Name and Version:

Makes showing up the header immediately possible

HoutarouOreki committed:
Update few links not working; Reasses sample playing; Slight renames

HoutarouOreki committed:
Fix reversed if/else statement causing inproper behavior for badges

HoutarouOreki committed:
Remove reduntant .base and .this

HoutarouOreki committed:
Add TextBadgePair test case

HoutarouOreki committed:
Remove unused variable

HoutarouOreki committed:
Slight reorder

HoutarouOreki committed:
Shorten MarginPaddings in ChangelogContentGroup

HoutarouOreki committed:
XML for TooltipIconButton.cs

HoutarouOreki committed:
Move breadcrumb to the right position

HoutarouOreki committed:
Prevent duplicated fetching for listing and builds:

- listing when already at it;
- builds by immediately disabling links to them (chevrons and links in listing)

HoutarouOreki committed:
Rename TestCaseChangelog to TestCaseChangelogOverlay

HoutarouOreki committed:
Add scrolling to the previous position when on "Back"

HoutarouOreki committed:
Improve scroll handling

HoutarouOreki committed:
Fetch listing only once; Reenable disabled links as they wont be regenerated

HoutarouOreki committed:
Non-breaking space for "by ..."

Doesn't work for now(?)

HoutarouOreki committed:
Make PR-link surrounding parenthesis white

HoutarouOreki committed:
Make Back action go to the top of listing before exiting the changelog screen

HoutarouOreki committed:
Hide graph when it's unavailable; Scroll change

HoutarouOreki committed:
Add links to GitHub

HoutarouOreki committed:
Add links to GitHub; Remove reduntant using directive and variable

HoutarouOreki committed:
Change initial chart colour

HoutarouOreki committed:
Bring StreamBadge visual proportions closer to the upstream design

HoutarouOreki committed:
Unify MarginPadding

HoutarouOreki committed:
Remove reduntant comment

HoutarouOreki committed:
Make ChangelogChart's child buffered

HoutarouOreki committed:
Move statement to a new line

HoutarouOreki committed:
Update comments

HoutarouOreki committed:
ClickableText changes colour on hover

HoutarouOreki committed:
Fix wrong conditional expression

HoutarouOreki committed:
Remove default value without use

HoutarouOreki committed:
Delete ClickableText class, use OsuHoverContainer instead

HoutarouOreki committed:
Merge remote-tracking branch 'upstream/master' into changelog-overlay

HoutarouOreki committed:
Plotting the chart 1

HoutarouOreki committed:
Plotting the chart 2

HoutarouOreki committed:
Add "ok" chart plotting; Add more keys to StreamColour .FromStreamName function

HoutarouOreki committed:
Resign ChangelogChart and disable its references

HoutarouOreki committed:
Merge remote-tracking branch 'upstream/master' into changelog-overlay

peppy commented on osu.Game/Graphics/UserInterface/LineBadge.cs:
```diff
@@ -0,0 +1,84 @@
+            }
+        }
+
+        /// <summary>
```
xmldoc should be on class, not ctor

peppy replied:
(this applies to every case where you have descriptions of classes on the ctor)

peppy commented on osu.Game/Online/API/Requests/Responses/APIChangelog.cs:
```diff
@@ -0,0 +1,120 @@
+
+namespace osu.Game.Online.API.Requests.Responses
+{
+    public class APIChangelog
```
shouldn't this be a `Build`?

peppy commented:
Currently you have:
```
APIxxx <- Model
xxx <- Drawable
```

We generally use:
```
xxx <- Model
Drawablexxx <- Drawable
```

instead

peppy commented:
Is this still going to get some attention?

HoutarouOreki committed:
Merge branch 'master' into changelog-overlay

HoutarouOreki committed:
Move xml docs from ctors onto classes

HoutarouOreki committed:
Remove unnecessary inheritance

HoutarouOreki committed:
Update comments

HoutarouOreki committed:
Merge branch 'master' into changelog-overlay

HoutarouOreki committed:
Update overrided functions to match their bases

HoutarouOreki committed:
Rename APIChangelog to APIChangelogBuild

HoutarouOreki committed:
Remove redundant usings

andy840119 commented:
@Hanamuke Markdown has been merged into osu-framework
So if you have time you can continuous working on it 👍

HoutarouOreki commented:
I tried a few days ago, but it seemed like Markdown Container doesn't
support Auto Size Axes, and that makes it kind of useless for this

On Sat, Jan 12, 2019, 3:53 PM 為什麼 <notifications@github.com wrote:

> @Hanamuke <https://github.com/Hanamuke> Markdown has been merged into
> osu-framework
> So if you have time you can continuous working on it 👍
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ppy/osu/pull/3074#issuecomment-453753561>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APgiZU2qswUHbrXcjjZTdzNAC_ovGMO_ks5vCfb4gaJpZM4VXAV5>
> .
>

andy840119 commented:
@HoutarouOreki is there any screenshot?
Or can you explain why it need screenshot?
Thanks : )

peppy committed:
Merge branch 'master' into changelog-overlay

peppy committed:
Resolve compile-time issues

peppy committed:
Fix deprecated calls and code styling (partly)

peppy committed:
Initial pass to make work with real API

peppy commented:
@HoutarouOreki I'm doing some work to bring this PR up-to-date and look at getting it merged. Just curious if you're still around, and also if you made further progress on the chart portion?

I'm going to tidy this up without the chart implementation for now, regarldess.

peppy committed:
Remove chart references for now

peppy committed:
Share web request between builds and streams

peppy committed:
Add toolbar button

peppy committed:
Update outdated licence headers

peppy committed:
Split out web response classes into own files

peppy committed:
Add JsonProperty hinting

peppy committed:
Adjust header image sizing

HoutarouOreki commented:
Having final exams for a few more days. As for the chart, there was practically no further progress than what's shown in the video in the original post.

peppy commented:
Thanks for the feedback. Could you also explain what the last checkbox (regex todo) refers to in the description?

peppy committed:
Merge branch 'add-new-colours' into HoutarouOreki-changelog-overlay

peppy committed:
Update ChangelogOverlay to use FullScreenOverlay

HoutarouOreki commented:
Ahh, that was supposed to help replace `TextSize` with the value from a comment (`// Web: xx,`) in bulk, by plugging it into Find&Replace in Visual Studio, but it doesn't work anymore since the way to change the font size has changed.

I was using different font sizes than these from the web to get a more 1:1 look on my monitor. For example, lazer's 28px font size is/was equivalent to 24px web's font size for me. Kept original sizes in comments to be able to later replace them without checking them through a browser.

![gif](https://user-images.githubusercontent.com/16261733/57672524-c4873a00-7618-11e9-95f7-41bb93d954a9.gif)

peppy commented:
i see, thanks!

peppy committed:
Merge branch 'fullscreen-overlay' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'master' into HoutarouOreki-changelog-overlay

peppy committed:
Tidy up colour references and texture assignment in ChangelogHeader

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

peppy committed:
Fix message display / html stripping

peppy committed:
Allow chaining of loadComponentSingleFile

peppy committed:
Hide other overlays when showing changelog

peppy committed:
Initial clean-up pass of breadcrumb and header

peppy committed:
Clean up events and states

peppy committed:
Remove unnecessary ReceivePositionalinputAt modification

peppy committed:
API request clean-up

peppy committed:
Fix another font usage case

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

peppy committed:
Bring up-to-date with master

peppy commented:
As a heads-up, the original code was such a mess I have considered multiple times starting from scratch. For now I will continue to work with this and refactor it, but there's still a long way to go.

I hope you can learn by watching your code be almost 100% rewritten from its original state.

Just a few things done wrong:
- Too many unnecessary bool flags and states
- Not using object oriented styles (every class has a `ShowListing` and `ShowBuild` instead of having two derived versions)
- Not using existing breadscrumb control (see `ScreenBreadcrumbControl`)
- No tests for smaller individual pieces (like a single changelog display)
- Not setup for dynamic compilation – I don't want to know how many times you rebuilt and restarted the app while working on this. Please see https://github.com/ppy/osu-framework/wiki/Development-and-Testing
- Rather than custom logic to show/hide listing/build and save scroll positions, it should probably be using a screen stack.
- Events and EventHandlers all over the place where they are not necessary.

HoutarouOreki commented:
Most of these were apparent to me at the point of considering MarkdownContainer, though I didn't have an exact precise idea of how they should be done correctly for most of them. Didn't want to bother myself too much with it because of my lessened amount of free time, considering that there was *so* much wrong. Your rewrite will definitely give some answers. Appreciated. Now that I think about it I should've closed the pr or edited the op to let others know I'm not gonna work on it for some time.

peppy committed:
Match header titles with web (pt size)

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

peppy committed:
Move resources to osu-resources

peppy committed:
Fix dynamic recompilation in TestSceneChangelogOverlay

peppy committed:
Convert BadgeDisplay to use bindable

peppy committed:
fixup! Convert BadgeDisplay to use bindable

peppy committed:
Remove custom scroll logic

peppy committed:
Split out ChangelogContent into two classes

peppy committed:
Use new colour palette

peppy committed:
Improve load and switch logic between views

peppy committed:
Only fetch after initial pop in

peppy committed:
Remove StreamColour class and implement locally

peppy committed:
Move new classes for now

peppy committed:
Update namespaces

peppy committed:
Merge remote-tracking branch 'upstream/master' into changelog-overlay

peppy committed:
Huge refactor pass focusing on ChangelogContent

peppy committed:
Move TooltipText to OsuClickableContainer

peppy committed:
Remove usage of TooltipIconButton completely

peppy committed:
Fix tests failing when not logged in

peppy committed:
Use OsuColour for profile overlay

peppy committed:
Move UserProfileOverlay's header into an abstract implementation

peppy committed:
Begin to consume abstract header implementation

peppy committed:
Remove unnecessary getters from ScreenTitle

peppy committed:
Allow custom icon specification in ScreenTitle

Not all icons are available in fonts so IconUsage alone is not enough to cover all scenarios.

peppy committed:
Set HeaderTabControl's default AccentColour to non-transparent

Avoids items disappearing if no accent colour is set.

peppy committed:
Update ChangelogHeader to work again with OverlayHeader

peppy committed:
Use bindables the whole way

peppy committed:
Move OverlayHeaderTabControl to correct namespace

peppy committed:
Delete breadcrumb implementation

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'move-tooltip-text' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'abstract-profile-header' into HoutarouOreki-changelog-overlay

peppy committed:
Fix incorrectly changed colour

peppy committed:
Remove redundant prefixes

peppy committed:
Merge branch 'hover-container-disabled-hover-effect' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'screen-title-custom-icon' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'fix-spritetext-usage' into HoutarouOreki-changelog-overlay

peppy committed:
Only propagate badge value changes if not the current UpdateStream

peppy committed:
Fix test scene

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

UselessToucan commented on osu.Game/Overlays/Changelog/ChangelogBuild.cs:
```diff
@@ -0,0 +1,168 @@
+            Children = new Drawable[]
+            {
+                CreateHeader(),
+                ChangelogEntries = new FillFlowContainer
```
I suggest to create a private method called `CreateList` or similar.

peppy replied:
why?

peppy replied:
you mean to move the parsing logic i haven't yet refactored? this PR is not ready for merge just yet.

peppy committed:
Remove unnecessary brackets

peppy committed:
Fix (and rename) ExpandingBar

peppy committed:
Merge branch 'expanding-bar' into HoutarouOreki-changelog-overlay

peppy committed:
Rename classes and fix back-to-front state

peppy committed:
Use TabControl instead of custom logic

peppy committed:
Fix badge state regressions from tab control usage

peppy committed:
Fix date string

peppy committed:
Move update streams inside header content

peppy committed:
Add support for osu! user links

peppy committed:
Centralise font specification

peppy committed:
Use linq instead of a temporary sorted list

peppy committed:
Cleanup pass

peppy committed:
Prepare changelog api requests to work when not logged in

peppy committed:
Further minor refactoring

peppy committed:
Merge remote-tracking branch 'upstream/master' into HoutarouOreki-changelog-overlay

peppy committed:
Merge branch 'update-framework' into HoutarouOreki-changelog-overlay

peppy committed:
Show changelog from new build notification

peppy committed:
Use TaskCompletionSource in a better manner

peppy committed:
Restore accidentally commented conditional

peppy committed:
Use a more friendly set method for tab control

peppy commented:
@HoutarouOreki This is all done now. working your code into a good state probably took 2-3x what it would have been writing from scratch, but i hope looking through my commits (and the approach to how i went about fixing pieces) is helpful to you.

Make sure to check our all the pull requests and fixes which spawned from this implementation (added to the original description of this PR).

peppy committed:
Merge branch 'master' into changelog-overlay

smoogipoo committed:
Merge remote-tracking branch 'Joehuu/master' into changelog-overlay

smoogipoo committed:
Fix bar not expanding/collapsing correctly

smoogipoo committed:
Fix overlay group + depth

smoogipoo committed:
Add comment + fix spinlocking

smoogipoo committed:
Fix size of release stream separator in listing

smoogipoo committed:
General refactoring

smoogipoo commented:
@peppy do a visual pass on it once more and check my recent commits.
